### PR TITLE
Fix prop_lfe_doc and lfedoc

### PR DESCRIPTION
**Pull request by yurrriq**
- Fix property tests for rearranged internal forms
- Move `example.lfe` to `dev/` and fix `bin/lfedoc`
  
  Also fix some mistakes in `example.lfe` and use a `define-function`
  once for good measure.

### CODE REVIEW

**Comment by yurrriq** [comment]
@rvirding, this needs review and can definitely be cleaned up, but I've squashed my commits and would imagine you get the general idea here.

File: src/lfe_doc.erl [hunk 1]
```diff
@@ -187,9 +187,13 @@ make_doc(Type, Name, Patterns, Doc, Line) ->
...
 
-string_to_binary(Str) when is_list(Str) ->
-    unicode:characters_to_binary(Str, utf8, utf8);
-string_to_binary(Bin) -> Bin.
+string_to_binary(Data) ->
+    case io_lib:char_list(Data) of
+        true ->
+            unicode:characters_to_binary(Data, utf8, utf8);
+        false when is_binary(Data) -> Data;
+        false -> <<"">>
```
Comment (yurrriq) [inline src/lfe_doc.erl]: This is especially bad.

**Comment by rvirding** [comment]
I don't think the `doc-string` should be optional in the `define-module`, `extend-module`, `define-function` and `define-macro` core forms. There they are a defined part of the form. It is for `defmodule`, `defun` and `defmacro` to handle the optional bit. If you actually use `extend-module` then I assume you know what you are doing.

If they were to become optional even there then I would prefer having a specific `(doc doc-string)` declaration, at least in `define-module` and `extend-module`. It is a bit more difficult in `define-function` and `define-macro` as they have a fixed number of elements.

**Comment by yurrriq** [comment]
~~I see. This breaks _lots_ of existing code then, which is fine, I suppose.~~

I'll need to update `ltest` ASAP for it's now invalid usage of `extend-module`.

I understand now. Can you update it s.t. the `doc-string` is optional at the right place in the code?

**Comment by yurrriq** [comment]
I can split out the test fixes.

**Comment by rvirding** [comment]
Test fixes definitely in. I had completely missed them, and lfedoc.

**Comment by yurrriq** [comment]
Thoughts on this now? I pretty much need to be able to do `(defmodule some-tests)` and either we support optional docstrings in `extend-module` or [ltest](https://github.com/lfex/ltest) needs an update.

With that said, I'd be happy to ditch the second commit in favor of some other solution.

**Comment by rvirding** [comment]
Fixing `defmodule` is no problem, that we will do. `extend-module` is easy to fix but I had viewed it as having the same args as `define-module` without the doc string. How much work is fixing ltest? :-)

**Comment by yurrriq** [comment]
There are two lines to change. :smile:

**Comment by rvirding** [comment]
I need to fix 2 lines for the defmodule issue so we are equal.

**Comment by rvirding** [comment]
Will the extend-module issue call major backward compatibility problems for ltest?

**Comment by rvirding** [comment]
One solution is to accept the "old" way, no first doc string, and issue a deprecated warning.

**Comment by yurrriq** [comment]
I'm curious to hear @oubiwann's opinion, but I think if we put an appropriate warning about LFE version on the ltest README we should be ok. It is a dramatic change though that breaks pretty much all tests.

**Comment by yurrriq** [comment]
When I get to my computer, I'll drop the second commit.

**Pull request merged**